### DISCUSSION TITLE
UI top bar -> listen for / display visualize active layer.

### DISCUSSION THREAD

**Discussion by molleweide** [kmonad/kmonad#420] [Q&A] [answered]
Hey guys,

What would be a "general" pattern for listening for currently active layer and then printing it to my computes top bar.
I only use macos but I am thinking something like displaying it in eg. polybar for linux, or I personally am trying to learn now how
to use ubersicht widget `simple-bar` for macos.

It would be pretty cool to show kmonad layer on macos with this: https://github.com/cmacrae/spacebar

My layers are growing and soon I will be having about twenty layers :D and so sometimes I lost in a swamp of testing layers and I felt that it will be a good thing to make it possible to do this.

This also relates a little bit to this question about vim leader (https://github.com/kmonad/kmonad/discussions/413). I am having a pretty nice tree structure design that will become more and more like vim. I am also going to try to implement a vim layer where I have backup vim commands for text fields that nothing works in.

In the end if I learn more about GUIs and stuff it would be super cool to have like a mini-map of the layers up in the corner or the top bar. so that one can never forget or lose track of mappings.

**Comment by MuhammedZakir** [comment, marked as answer]
A simple way is to wrap your layer switching buttons in `tap-macro` to execute a shell command before switching layers.

```clojure
(tap-macro (cmd-button "notify-bar :layer symbols") (layer-next symbols))
```

Because of how long it takes to execute a shell command and get the bar to display it, this won't be useful if layers are changed rapidly.

**Reply by molleweide** [reply]
Wow interesting! Im gonna and see if I can get this to work.

**Reply by glostis** [reply]
For future reference to anyone who might be interested by the solution for polybar specifically, I have found that using the [IPC module](https://github.com/polybar/polybar/wiki/Module:-ipc) works well.

Below a minimal example:
```dosini
; polybar config: an empty `custom/ipc` module called `kmonad`
[module/kmonad]
type = custom/ipc
```

```bash
# Use the following snippet to print the string `symbols` in the polybar module
(tap-macro (cmd-buttonK "polybar-msg action kmonad send 'symbols'") (layer-next symbols))
```

**Reply by molleweide** [reply]
thanks for adding this!
